pkg/flow/worker/worker_pool.go

```diff
@@ -0,0 +1,124 @@
+package worker
+
+import (
+	"fmt"
+	"hash/fnv"
+	"math/rand"
+	"runtime"
+	"sync"
+)
+
+type Pool interface {
+	// Stop stops the worker pool. It does not wait to drain any internal queues. It must only be called once.
+	Stop()
+	// Submit submits a function to be executed by the worker pool on a random worker. Error is returned if the pool
+	// is unable to accept extra work.
+	Submit(func()) error
+	// SubmitWithKey submits a function to be executed by the worker pool, ensuring that only one
+	// job with given key can be queued at the time. Adding a job with a key that is already queued is a no-op.
+	// Error is returned if the pool is unable to accept extra work.
+	SubmitWithKey(string, func()) error
+	// QueueSize returns the number of tasks currently queued.
+	QueueSize() int
+}
+
+// shardedWorkerPool is a Pool that distributes work across a fixed number of workers, using a hash of the key to
+// determine which worker to use. This, to an extent, defends the pool from a slow task hogging all the workers.
+type shardedWorkerPool struct {
+	workersCount int
+	workQueues   []chan func()
+	quit         chan struct{}
+
+	lock sync.Mutex
+	set  map[string]struct{}
+}
+
+var _ Pool = (*shardedWorkerPool)(nil)
+
+func NewDefaultWorkerPool() Pool {
+	return NewShardedWorkerPool(runtime.NumCPU(), 100)
+}
+
+// NewShardedWorkerPool creates a new worker pool with the given number of workers and queue size for each worker.
+// The queued tasks are sharded across the workers using a hash of the key. The pool is automatically started and
+// ready to accept work. To prevent resource leak, Stop() must be called when the pool is no longer needed.
+func NewShardedWorkerPool(workersCount int, queueSize int) Pool {
+	if workersCount <= 0 {
+		panic(fmt.Sprintf("workersCount must be positive, got %d", workersCount))
```

Comment:
Can we return an error instead of panicking?

Reply:
This would cascade to many calling functions, so I would prefer not to do that. This should never happen anyway unless we make a coding error. Note that this is not that unusual as it seems, the standard library does it too, e.g. creating a channel with negative size will panic too.